src/Migrations/HSLMigration.php

```diff
@@ -0,0 +1,711 @@
+        'name' => 'search_last',
+        'replace_false_with_null' => true,
+      ),
+      'implode' => shape(
+        'ns' => HslNamespace::STR,
+        'name' => 'join',
+        'argument_order' => vec[1, 0],
+      ),
+      'join' => shape(
+        'ns' => HslNamespace::STR,
+        'name' => 'join',
+        'argument_order' => vec[1, 0],
+      ),
+      'substr_replace' => shape(
+        'ns' => HslNamespace::STR,
+        'name' => 'splice',
+        'has_overrides' => true,
+      ),
+      'substr' => shape(
+        'ns' => HslNamespace::STR,
+        'name' => 'slice',
+        'has_overrides' => true,
+      ),
+      'str_repeat' => shape('ns' => HslNamespace::STR, 'name' => 'repeat'),
+      'trim' => shape('ns' => HslNamespace::STR, 'name' => 'trim'),
+      'ltrim' => shape('ns' => HslNamespace::STR, 'name' => 'trim_left'),
+      'rtrim' => shape('ns' => HslNamespace::STR, 'name' => 'trim_right'),
+      'strlen' => shape('ns' => HslNamespace::STR, 'name' => 'length'),
+      'sprintf' => shape('ns' => HslNamespace::STR, 'name' => 'format'),
+      'str_split' => shape('ns' => HslNamespace::STR, 'name' => 'chunk'),
+      'strcmp' => shape('ns' => HslNamespace::STR, 'name' => 'compare'),
+      'strcasecmp' => shape('ns' => HslNamespace::STR, 'name' => 'compare_ci'),
+      'number_format' =>
+        shape('ns' => HslNamespace::STR, 'name' => 'format_number'),
+
+      // Math functions
+      'round' => shape(
+        'ns' => HslNamespace::MATH,
+        'name' => 'round',
+        'has_overrides' => true,
+      ),
+      'ceil' => shape('ns' => HslNamespace::MATH, 'name' => 'ceil'),
+      'floor' => shape('ns' => HslNamespace::MATH, 'name' => 'floor'),
+      'array_sum' => shape('ns' => HslNamespace::MATH, 'name' => 'sum'),
+      'intdiv' => shape('ns' => HslNamespace::MATH, 'name' => 'int_div'),
+      'exp' => shape('ns' => HslNamespace::MATH, 'name' => 'exp'),
+      'abs' => shape('ns' => HslNamespace::MATH, 'name' => 'abs'),
+      'base_convert' =>
+        shape('ns' => HslNamespace::MATH, 'name' => 'base_convert'),
+      'cos' => shape('ns' => HslNamespace::MATH, 'name' => 'cos'),
+      'sin' => shape('ns' => HslNamespace::MATH, 'name' => 'sin'),
+      'tan' => shape('ns' => HslNamespace::MATH, 'name' => 'tan'),
+      'sqrt' => shape('ns' => HslNamespace::MATH, 'name' => 'sqrt'),
+      'log' => shape('ns' => HslNamespace::MATH, 'name' => 'log'),
+      'min' => shape(
+        'ns' => HslNamespace::MATH,
+        'name' => 'min',
+        'has_overrides' => true,
+      ),
+      'max' => shape(
+        'ns' => HslNamespace::MATH,
+        'name' => 'max',
+        'has_overrides' => true,
+      ),
+
+      // Container functions
+      'count' => shape(
+        'ns' => HslNamespace::C,
+        'name' => 'count',
+        'argument_order' => vec[0],
+      ),
+    ];
+
+
+  <<__Override>>
+  public function migrateFile(string $path, EditableNode $root): EditableNode {
```

Comment:
heh, I should change this to be an async function in the next version. THis is fine :)